File: #137_Single_Number_II/Solution2.py
class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def singleNumber(self, nums):
    	x1 = 0
    	x2 = 0
    	mask = 0
    	for i in nums:
    		x2 ^= x1 & i
    		x1 ^= i
    		mask = ~(x2 & x1) # k = 11
    		x2 &= mask
    		x1 &= mask
    	return x2 # p = 10

sol = Solution()
print sol.singleNumber([1, 1, 1, 3, 3, 3, 4, 4])

File: #220_Contains_Duplicate_III/Solution2.java
import java.util.*;

public class Solution2 {
    public boolean containsNearbyAlmostDuplicate(int[] nums, int k, int t) {
        if (nums == null || nums.length == 0) return false;
        if (k <= 0 || t < 0) return false;
        TreeSet<Integer> set = new TreeSet<Integer>();
        for (int i = 0; i < nums.length; i++) {
            Integer floor = set.floor(nums[i] + t);
            Integer ceil = set.ceiling(nums[i] - t);
            if (floor != null && floor >= nums[i]) return true;
            if (ceil != null && ceil <= nums[i]) return true;
            set.add(nums[i]);
            if (i >= k) {
                set.remove(nums[i - k]);
            }
        }
        return false;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] nums = {1, 2, 3, 4, 5, 6, Integer.MAX_VALUE};
        System.out.println(sol.containsNearbyAlmostDuplicate(nums, 2, 1));
    }
}